### ISSUE TITLE
False negative when inferring paramspec for a class constructor

### ISSUE THREAD

**Issue by Wizzerinus** [astral-sh/ty#2587]
### Summary

The following playground link emits no errors:

```py
from typing import Generic, ParamSpec, Protocol

P = ParamSpec("P")


class ParentClass:
    pass


class MyProto(Protocol, Generic[P]):
    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> ParentClass: ...


class MyType(ParentClass):
    def __init__(self, x: int):
        pass


def create(p: MyProto[P], *args: P.args, **kwargs: P.kwargs):
    pass


create(MyType)
```

https://play.ty.dev/8273c9c2-da7f-4432-982d-71bba5c60dbe

However, this code is not sound.

To typecheck this method, P must be inferred empty:

```py
def create(p: MyProto[P], *args: P.args, **kwargs: P.kwargs):
    pass


create(MyType)
```

Hence, the function must accept no parameters except self:
```py
class MyProto(Protocol, Generic[P]):
    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> ParentClass: ...
```

However, `MyType` is defined as follows:
```py
class MyType(ParentClass):
    def __init__(self, x: int):
        pass
```

So it cannot be called with no parameters. (This is not about `self` being considered a parameter - the result is the same regardless of how many parameters `__init__` wants here.)

### Version

0.0.13

**Comment by Wizzerinus** [comment]
There's also an opposite example, emitting a false positive:
```py
from typing import Generic, ParamSpec, Protocol

P = ParamSpec("P")


def my_factory(arg: str) -> int:  # or have this a class constructor, doesn't matter
    return 0


class Factory(Protocol, Generic[P]):
    def __call__(self, arg: str, *args: P.args, **kwargs: P.kwargs) -> int: ...


def create(ctr: Factory[P], *args: P.args, **kwargs: P.kwargs):
    ctr("", *args, **kwargs)


create(my_factory)
```

https://play.ty.dev/0d16fa1c-ccd2-4ffc-af0a-6cdc466b1036

It seems entirely dependent on whether the protocol has a parameter outside of the ParamSpec. Maybe not constructor-related.

**Comment by AlexWaygood** [comment]
(nit: could you possibly paste the full playground content into the issue body as well as pasting a link to the playground, when you make an issue? It makes it much easier for us to see at a glance what the issue is about, without us having to click through to open the playground link in another tab :-)

**Comment by Wizzerinus** [comment]
done, sorry:)

**Comment by carljm** [comment]
Thanks for the report! I believe this is a known issue, and has the same root cause as https://github.com/astral-sh/ty/issues/1970 -- but I'll leave this open, as it has nice self-contained repro examples.
